dbus/mako.c

```diff
@@ -14,82 +14,84 @@
 static const char *service_path = "/fr/emersion/Mako";
 static const char *service_interface = "fr.emersion.Mako";
 
-static int handle_dismiss_all_notifications(sd_bus_message *msg, void *data,
+static int handle_dismiss(sd_bus_message *msg, void *data,
 		sd_bus_error *ret_error) {
 	struct mako_state *state = data;
 
-	close_all_notifications(state, MAKO_NOTIFICATION_CLOSE_DISMISSED);
-	struct mako_surface *surface;
-
-	wl_list_for_each(surface, &state->surfaces, link) {
-		set_dirty(surface);
+	if (wl_list_empty(&state->notifications)) {
+		goto done;
 	}
 
-	return sd_bus_reply_method_return(msg, "");
-}
-
-static int handle_dismiss_group_notifications(sd_bus_message *msg, void *data,
-		sd_bus_error *ret_error) {
-	struct mako_state *state = data;
+	uint32_t id = 0;
+	int group = 0;
+	int no_history = 0;
+	int all = 0;
 
-	if (wl_list_empty(&state->notifications)) {
-		goto done;
+	int ret = sd_bus_message_enter_container(msg, 'a', "{sv}");
+	if (ret < 0) {
+		return ret;
 	}
 
-	struct mako_notification *notif =
-		wl_container_of(state->notifications.next, notif, link);
+	while (true) {
+		ret = sd_bus_message_enter_container(msg, 'e', "sv");
+		if (ret < 0) {
+			return ret;
+		} else if (ret == 0) {
+			break;
+		}
 
-	struct mako_surface *surface = notif->surface;
-	close_group_notifications(notif, MAKO_NOTIFICATION_CLOSE_DISMISSED);
-	set_dirty(surface);
+		const char *key = NULL;
+		ret = sd_bus_message_read(msg, "s", &key);
+		if (ret < 0) {
+			return ret;
+		}
 
-done:
-	return sd_bus_reply_method_return(msg, "");
-}
+		if (strcmp(key, "id") == 0) {
+			ret = sd_bus_message_read(msg, "v", "u", &id);
+		} else if (strcmp(key, "group") == 0) {
+			ret = sd_bus_message_read(msg, "v", "b", &group);
+		} else if (strcmp(key, "no-history") == 0) {
```

Comment:
Instead of using negative options, I'd prefer to define a positive option and default it to true:

```c
int history = 1;
if (strcmp(key, "history") == 0) {
	ret = sd_bus_message_read(msg, "v", "b", &history);
}
```

Reply:
Should it still stay as `--no-history` in makoctl or do you prefer it to be a positive option there as well?
I used a negative option to be consistent with the `--no-history` option for the `dismiss` action in the config.

Reply:
Yeah, I think for CLI arguments we can stick to negative options, because it's more consistent with the status quo and because boolean CLI options don't take a value.